### PR TITLE
feat: relic data, wfm names, type hardening

### LINKED ISSUE #365: :rocket: Incorporate warframe-relic-data
**Is your enhancement request related to a problem? Please describe.**
TitaniaProject/warframe-relic-data has made a lot of potentially helpful data connections for relics, would be worth integrating into our relic data

**Describe the upgrade you'd like**
incorporate TitaniaProject/warframe-relic-data optimizations into our style and make available as a type

**Describe alternatives you've considered**
just parsing some of our own data and correlating warframe market ids

**Additional context**
also logged issue here: https://github.com/TitaniaProject/warframe-relic-data/issues/7 to see if they'd be interested in collabing on it

### LINKED ISSUE #368: Inconsistent or incorrect data for exalted weapons
### Describe the bug

Exalted weapons have inconsistent or incorrect damagePerShot, types, tags and others.

### NodeJS Version

v16.16.0

### npm Version

v8.16.0

### warframe-items Version

v1.1253.2595

### Access Method

npm Package (Default)

### What happened?

- All exalted weapons, including Necramech ones, and Garuda Talons are included in Misc.json which might be intentional but is very inconvenient when using all weapon data. Fetching over 600 items to use only 22 is pretty inefficient. But that might only be a me problem.
- Ironbride has the "Rifle" `type`.
- Only Artemis Bow and Arquebex have the `type` "Exalted Weapon" while the others have "Warframe". Including Artemis Bow Prime.
- Garuda Prime Talons have the value "Prime" while all the others have "Tenno" in `tags`. Except Ironbride and Arquebex which correctly have "Entrati".
- Artemis Bow, Artemis Bow Prime, Balefire Charger have 0 `damagePerShot` and `totalDamage`.
- Desert Wind, Dex Pixia, Dex Pixia Prime, Diwata, Diwata Prime, Exalted Blade, Exalted Prime Blade, Exalted Umbra Blade, Ironbride, Iron Staff, Iron Staff Prime, Shadow Claws, Valkyr Talons and Valkyr Prime `damagePerShot` values are inconsistent with the `attacks` values.
- Desert Wind, Ironbride, Valkyr Talons and Valkyr Prime Talons have an incorrect `procChance` value.
- Dex Pixia and Dex Pixia Prime have an incorrect `reloadTime` value. Arquebex, Balefire Charger, Regulators and Regulators Prime also have a `reloadTime` value despite never reloading but that's not a real issue.

Note that there might be more. I only provided the ones I currently work with.

### PR DESCRIPTION
**Pull request by TobiTenno**
### What did you fix? (provide a description or issue closes statement)
closes #365 
closes #368 

---

### Reproduction steps
1. All the types
2. Split by category/type/etc
3. ...?
4. Profit

---

### Evidence/screenshot/link to line

You can see my fixes mostly in index.d.ts

### Considerations
- Does this contain a new dependency? **[Yes]**
- Does this introduce opinionated data formatting or manual data entry? **[Yes]**
- Does this pr include the updated data files in a separate commit that can be reverted for a clean code-only pr? **[Yes]**
- Have I run the linter through `npm run lint`? **[Yes]**
- Is is a bug fix, feature request, or enhancement? **[Feature]**

### CODE REVIEW

**Comment by TobiTenno** [comment]
@AyAyEm i'd really appreciate your thoughts since you've done more ts than i have by far

**Comment by TobiTenno** [comment]
whoops, i think i commented out the stuff for loading i18n values

File: build/parser.js
Comment (AyAyEm) [inline build/parser.js:940]: All this set creation is for future proffing? Because there's no other need for it, the property `name`, in the generated data from titania project relics, are all unique.

**Comment by AyAyEm** [comment]
So far i've gathered some data comparing between the **WFCD** relic data and **TP** (titania project) relic data.

- All the relics in **TP** are in **WFCD**.
- **TP** relics doesn't contain the variants (Intact, ..., Radiant) in their names, instead it's in their properties
- The requiem casing naming for the relics is different for the requiem variant.
In **TP** they are: 
`'Requiem I', 'Requiem II', 'Requiem III',  'Requiem IV'`
In **WFCD** they are:
`'Requiem I', 'Requiem Ii', 'Requiem Iii',  'Requiem Iv'`

I'll update this as i find out more things to be aware when working with this PR.

File: build/parser.js
Comment (TobiTenno) [reply build/parser.js:940]: Yeah, just in case something gets in that shouldn't be in and makes it non-unique.

**Comment by TobiTenno** [comment]
> So far, i've gathered some data comparing the **WFCD** relics and **TP** ([Titania Project](https://github.com/TitaniaProject/warframe-relic-data)) relics.
> 
> - All the relics in **TP** are in **WFCD**.
> - **TP** relics doesn't contain the variants (Intact, ..., Radiant) in their names, instead it's in their properties
> - The requiem casing naming for the relics is different for the requiem variant.
> In **TP** they are: 
> `'Requiem I', 'Requiem II', 'Requiem III',  'Requiem IV'`
> In **WFCD** they are:
> `'Requiem I', 'Requiem Ii', 'Requiem Iii',  'Requiem Iv'`
> 
> I'll update this as i find out more things to be aware when working with this PR.

Good to know. Probably need to detect and fix those...

File: index.d.ts
Comment (AyAyEm) [inline index.d.ts:232]: I think these long union types are better outside the interface, for aesthetics and easy to reuse.
```suggestion
    type ModType = 
        'Railjack' |
        'Necramech' |
        'Warframe' |
        'Secondary' |
        'Melee' |
        'Companion' |
        'Primary' |
        'K-Drive' |
        'Riven' |
        'Archwing' |
        'Arch-Melee' |
        'Arch-Gun' |
        'Shotgun' |
        'Creature' |
        'Stance' |
        'Parazon' |
        'Transmutation' |
        'Peculiar';
    interface Mod extends MinimalItem, WikiaItem, Droppable {
        baseDrain: number;
        category: 'Mods';
        compatName?: string;
        fusionLimit: number;
        isAugment?: boolean;
        isPrime: false;
        levelStats?: LevelStat[];
        polarity: Polarity;
        transmutable?: boolean;
        /** @deprecated probably replaced by Mod#isExilus */
        isUtility?: boolean;
        modSet?: UniqueName;
        isExilus?: boolean;
        modSetValues?: number[];
        type: `${ModType} Mod`;
    }
```

File: index.d.ts
Comment (AyAyEm) [inline index.d.ts:546]: This would also be for future proffing? Because all relics have a warframe market link.

**Review by AyAyEm — COMMENTED** [review]
Great work.

File: index.d.ts
Comment (TobiTenno) [reply index.d.ts:548]: ah, unfortunately Forma (any TitaniaRelicRewardItem w/ name: 'Forma') do not.

**Comment by TobiTenno** [comment]
> > So far, i've gathered some data comparing the **WFCD** relics and **TP** ([Titania Project](https://github.com/TitaniaProject/warframe-relic-data)) relics.
> > 
> > * All the relics in **TP** are in **WFCD**.
> > * **TP** relics doesn't contain the variants (Intact, ..., Radiant) in their names, instead it's in their properties
> > * The requiem casing naming for the relics is different for the requiem variant.
> >   In **TP** they are:
> >   `'Requiem I', 'Requiem II', 'Requiem III',  'Requiem IV'`
> >   In **WFCD** they are:
> >   `'Requiem I', 'Requiem Ii', 'Requiem Iii',  'Requiem Iv'`
> > 
> > I'll update this as i find out more things to be aware when working with this PR.
> 
> Good to know. Probably need to detect and fix those...

i guess i could just toLowercase before checking equality. simple enough matching. i wanna fix requiem names, but i don't wanna do them in this PR :rofl:

File: index.d.ts
Comment (Soundofdarkness) [reply index.d.ts:548]: So far all items that contain "Forma", "Kuva", "Exilus" or "Riven" won't have any wfm item info, since they aren't tradable afaik. But I can see how that is more annoying to work with. I could add them later today if you want

**Comment by Soundofdarkness** [comment]
Im not sure what would be more accurate, I always assumed that roman numerals would be upper case, but I can change that if you would prefer that. So far I just took the naming from the droptables iirc

File: index.d.ts
Comment (TobiTenno) [reply index.d.ts:548]: Ultimately, as long as they're marked nullable, i dunno that it's a problem? It'd be nice if they weren't, but that's just reality, lol

Comment (Soundofdarkness) [reply index.d.ts:548]: Its not a problem per say, just one more null check, but otherwise the id and url would be null i guess 😄

Comment (TobiTenno) [reply index.d.ts:548]: Yeah, hopefully people are wisely using existence chaining.

**Comment by TobiTenno** [comment]
> Im not sure what would be more accurate, I always assumed that roman numerals would be upper case, but I can change that if you would prefer that. So far I just took the naming from the droptables iirc

nono, yours is right, ours is formatted.... badly

File: index.d.ts
Comment (AyAyEm) [reply index.d.ts:548]: Actually it's the relic warframe market data, i mistook by its reward. Having a null or undefined to represent an optional value is totally fine.

**Comment by Soundofdarkness** [comment]
Alright, and well, since it worked with formatting before requiems I guess I would have done the same if something needed reformatting 😄

File: index.d.ts
Comment (AyAyEm) [inline index.d.ts:287]: Is it necessary to have an union with null here? As every relic is tradable and then would have a link to warframeMarket, but considering the case of a new releases item it takes some time for it to be in warframe market.

Comment (TobiTenno) [reply index.d.ts:287]: That seems like you answered your own question....

**Comment by TobiTenno** [comment]
> Alright, and well, since it worked with formatting before requiems I guess I would have done the same if something needed reformatting 😄 I'm not sure if that could just be fixed by checking for "requiem" and "relic" in the name and then not applying title case or if that would break something else I didn't see in the data. But in case that would be an acceptable fix I could make a PR for that if that would help.
> 

Feel free, I'm gonna wait until relic data is ready to get into a module to merge this anyway, cause I have a feeling the action cache doesn't handle git installs very nicely.

**Comment by Soundofdarkness** [comment]
Alright, PR is ready for that change(#375). I hope that the PR is correct like this, so far I'm not used to working on other repositories. And for the module part of waframe-relic-data I guess the first step would probably be to migrate the repo into this organization right ?

**Comment by TobiTenno** [comment]
> Alright, PR is ready for that change(#375). I hope that the PR is correct like this, so far I'm not used to working on other repositories. And for the module part of waframe-relic-data I guess the first step would probably be to migrate the repo into this organization right ?

sure! let's coordinate on discord

**Comment by AyAyEm** [comment]
@Soundofdarkness 
All the vault date properties, of the relics, is currently an empty string, is it a planned feature?

**Comment by Soundofdarkness** [comment]
Unfortunately I don't think I can get that date. Somehow I thought I had it when I was writing the schema, but completely forgot about it later on, sorry. Unless you have that data somewhere I think I can remove that property.

**Comment by AyAyEm** [comment]
> Unfortunately I don't think I can get that date. Somehow I thought I had it when I was writing the schema, but completely forgot about it later on, sorry. Unless you have that data somewhere I think I can remove that property.

@TobiTenno
What do you think?

**Comment by TobiTenno** [comment]
Yeah, let's remove it and look at a way to get it in item itself.

File: index.d.ts
Comment (AyAyEm) [inline index.d.ts:546]: Should it extends `MinimalItem`, by having an unique name? If so i think it would be `warframe-items` responsibility to add it. 
```suggestion
    interface TitaniaRelicRewardItem extends MinimalItem {
```

Comment (TobiTenno) [reply index.d.ts:546]: It doesn't have a uniqueName, I don't think...

Comment (AyAyEm) [reply index.d.ts:546]: > It doesn't have a uniqueName, I don't think...

Yeah that's why it would be `warframe-items` responsibility to add it. By checking the for an item with the same name and getting it's uniqueName, and probably we would have to make some other guesses to narrow it down.

Comment (TobiTenno) [reply index.d.ts:546]: Ah, I see what you mean. We could do that.

Comment (AyAyEm) [reply index.d.ts:546]: Should it be included in this PR or in another one?

Comment (TobiTenno) [reply index.d.ts:546]: Before I merge this, I'm doing some ci work on relics, so we can probable do that on there shortly. Ideally we'll be able to drop our extra types

Comment (TobiTenno) [reply index.d.ts:546]: looks like we can't drop them just yet, and we don't have unique names on them (yet)... do we want that before initial support?

Comment (AyAyEm) [reply index.d.ts:546]: > looks like we can't drop them just yet, and we don't have unique names on them (yet)... do we want that before initial support?

No need, it's just a nice to have in case we want to find more information about some item.

Comment (TobiTenno) [reply index.d.ts:546]: unfortunately rebasing has some ci issues, so i'll get those sorted, probably rebase again, then poke again for re-review

**Review by AyAyEm — APPROVED** [review]

**Pull request merged**
